govcd/api_vcd_test.go

```diff
@@ -150,57 +181,144 @@ func (vcd *TestVCD) SetUpSuite(check *C) {
+	if vcdClient == nil || err != nil {
 		panic(err)
 	}
 	vcd.client = vcdClient
+	if config.Provider.SysOrg != "System" {
+		fmt.Printf("Skipping OrgAdmin tests\n")
+		vcd.skipAdminTests = true
+	}
 	// org and vdc are the test org and vdc that is used in all other test cases
 	err = vcd.client.Authenticate(config.Provider.User, config.Provider.Password, config.Provider.SysOrg)
 	if err != nil {
 		panic(err)
 	}
 	// set org
 	vcd.org, err = GetOrgByName(vcd.client, config.VCD.Org)
-	if err != nil {
+	if err != nil || vcd.org == (Org{}) {
 		panic(err)
 	}
 	// set vdc
 	vcd.vdc, err = vcd.org.GetVdcByName(config.VCD.Vdc)
-	if err != nil {
+	if err != nil || vcd.vdc == (Vdc{}) {
 		panic(err)
 	}
 	// creates a new VApp for vapp tests
 	if config.VCD.Network != "" && config.VCD.StorageProfile.SP1 != "" &&
 		config.VCD.Catalog.Name != "" && config.VCD.Catalog.Catalogitem != "" {
-		vcd.vapp, err = vcd.createTestVapp(testVappName)
-		if err != nil {
+		vcd.vapp, err = vcd.createTestVapp(TestSetUpSuite)
+		// If no vApp is created, we skip all vApp tests
+		if vcd.vapp == (VApp{}) || err != nil {
 			fmt.Printf("%v", err)
 			vcd.skipVappTests = true
 		}
+		// After a successful creation, the vAPp is added to the cleanup list
+		AddToCleanupList(TestSetUpSuite, "vapp", "", "SetUpSuite")
 	} else {
 		vcd.skipVappTests = true
 		fmt.Printf("Skipping all vapp tests because one of the following wasn't given: Network, StorageProfile, Catalog, Catalogitem")
 	}
-
 }
 
-func (vcd *TestVCD) TearDownSuite(check *C) {
-	if vcd.skipVappTests {
-		check.Skip("Vapp tests skipped, no vapp to be deleted")
+// Shows the detail of cleanup operations only if the relevant verbosity
+// has been enabled
+func (vcd *TestVCD) infoCleanup(format string, args ...interface{}) {
+	if vcd.config.Logging.VerboseCleanup {
+		fmt.Printf(format, args...)
 	}
-	err := vcd.vapp.Refresh()
-	if err != nil {
-		panic(err)
-	}
-	task, _ := vcd.vapp.Undeploy()
-	_ = task.WaitTaskCompletion()
-	task, err = vcd.vapp.Delete()
-	if err != nil {
-		panic(err)
+}
+
+// Removes leftover entities that may still exist after failed tests
+// or the ones that were explicitly created for several tests and
+// were relying on this procedure to clean up at the end.
+func (vcd *TestVCD) removeLeftoverEntities(entity CleanupEntity) {
+	var introMsg string = "removeLeftoverEntries: [INFO] Attempting cleanup of %s '%s' instantiated by %s\n"
+	var notFoundMsg string = "removeLeftoverEntries: [INFO] No action for %s '%s'\n"
+	var removedMsg string = "removeLeftoverEntries: [INFO] Removed %s '%s' created by %s\n"
+	var notDeletedMsg string = "removeLeftoverEntries: [ERROR] Error deleting %s '%s': %s\n"
+	// NOTE: this is a cleanup function that should continue even if errors are found.
+	// For this reason, the [ERROR] messages won't be followed by a program termination
+	vcd.infoCleanup(introMsg, entity.Name, entity.EntityType, entity.CreatedBy)
+	switch entity.EntityType {
+	case "vapp":
```

Comment:
I am only concern here is that method soon will be huge, cause all the cases will have they own code. So general method readability what it does will degrade.

Reply:
Great